src/tools/pub/automated-publishing.md

```diff
@@ -0,0 +1,719 @@
+---
+title: Automate publishing packages to pub.dev
+description: Publish Dart packages to pub.dev directly from GitHub Actions.
+---
+
+You can automate publishing package updates to pub.dev using
+GitHub Actions.
+
+{{site.alert.note}}
+  Today, you can only automate publishing of existing packages.
+  To create a new package, publish the first version using
+  `dart pub publish`.
+{{site.alert.end}}
+
+This automation requires authenticating to `pub.dev` using either
+[GitHub-signed OpenID Connect tokens][1] or a [GCP Service Account][2].
+
+With these methods of automated publishing, you don't need to store
+long-lived secrets. The `dart pub` client authenticates using temporary
+tokens that either GitHub Actions or Google Cloud IAM have signed.
+
+
+## Enable publishing packages using GitHub Actions
+
+To publish from [GitHub Actions](https://github.com/features/actions),
+you must:
+
+1. Enable automated publishing for the package on `pub.dev`.
+   You need to provide the following information in `pub.dev`:
+   * Your package's GitHub _repository_ from which you publish the
+     package.
+   * A _`tag-pattern`_ that git tags must match to permit publishing.
+1. Create a GitHub Action workflow for publishing in GitHub.
+1. Create a `pubspec.yaml` file in your repository.
+   This file requires a `version` key with a value that matches the
+   `tag-pattern`.
+1. Create a git tag for your latest commit.
+   This tag also must match the `tag-pattern`.
+
+{{site.alert.note}}
+  GitHub Actions can publish to `pub.dev` only when a push of a git tag
+  to GitHub triggers the Action _workflow_.
+  If a GitHub Actions triggers publishing without a tag, `pub.dev`
+  rejects it.
+  This ensures that events that don't trigger publish result in
+  publishing new versions.
+{{site.alert.end}}
+
+
+## Configure automated publishing from GitHub Actions on pub.dev
```

Comment:
Perhaps an outline of this document could be:
 * Introduction: stating you can publish from github actions, gcb, or anywhere with service accounts.
 * Section: Publishing packages using Github Actions
    * Outline of steps required to do this:
       * Enable automated publishing on pub.dev
       * Create a workflow file
       * Trigger publishing
    * Subsection: Configure automated publishing from GitHub Actions on pub.dev
       * How to do the things that has to do be done on pub.dev
       * How `tag-pattern` works, what it is...
    * Subsection: Configure a GitHub Action workflow to publish to pub.dev
      * How to create the workflow file `.github/workflows/publish.yml`
      * How to create a custom workflow, if needed.
    * Subsection: Triggering automated publishing from Github Actions
      * How to update the version in `pubspec.yaml` (or just stating that you need to)
      * How to create and push a tag
      * Where to find the workflow logs
      * How things show up in audit-log on pub.dev
    * Subsection: Troubleshooting publishing from GitHub Actions
      * Check workflow logs
      * Check pattern triggering the workflow
      * Check tag-pattern on pub.dev
    * Hardening security when publishing from Github Actions
      * Why you might want to harden security (because anyone with push access to repo, can trigger publish)
      * Subsubsection: Hardening security with tag protection rules on GitHub
        * How to use tag protection rules
        * Limitations of tag protection rules
        * Suggestiong to use deployment environments (next section)
      * Subsubsection: Hardening security with GitHub Deployment Environments
        * How to require a deployment environment on pub.dev
        * How to create deployment environment on Github
        * How to alter your workflow
        * FACT: environment is reflected in the OIDC token and enforced by pub.dev; modifying workflow files won't work around it.
        * Suggestion to use "environment protection rules" on github, which supports "required reviewers"
        * Screenshot: github action pending review

I think splitting "Configure a GitHub Action workflow to publish to pub.dev" into multiple subsections is confusing. And I think a subsection in the middle on updating "pubspec.yaml" is confusing.

I'm not sure, we need "Hardening security when publishing from Github Actions" as this creates subsubsections for the two ways of hardening. And we could just put the WHY under the first hardening section.

I'm also not sure we need a separate troubleshooting section, we could put that under the triggering section. But maybe that's also fine and useful.

And other tweak on the outline could be:
   * Introduction: stating you can publish from github actions, gcb, or anywhere with service accounts.
 * Section: Publishing packages using Github Actions
    * Outline of steps required to do this:
       * Enable automated publishing on pub.dev
       * Create a workflow file
       * Trigger publishing
    * Subsection: Configure automated publishing from GitHub Actions on pub.dev
       * How to do the things that has to do be done on pub.dev
       * How `tag-pattern` works, what it is...
    * Subsection: Configure a GitHub Action workflow to publish to pub.dev
      * How to create the workflow file `.github/workflows/publish.yml`
      * How to create a custom workflow, if needed.
    * Subsection: Triggering automated publishing from Github Actions
      * How to update the version in `pubspec.yaml` (or just stating that you need to)
      * How to create and push a tag
      * Where to find the workflow logs
      * How to troubleshoot: check workflow logs, check pattern in workflow, check pattern on pub.dev
      * How things show up in audit-log on pub.dev
    * Subsection: Hardening security with tag protection rules on GitHub
      * Why you might want to harden security (because anyone with push access to repo, can trigger publish)
      * How to use tag protection rules
      * Limitations of tag protection rules
      * Suggestiong to use deployment environments (next section)
    * Subsection: Hardening security with GitHub Deployment Environments
      * How to require a deployment environment on pub.dev
      * How to create deployment environment on Github
      * How to alter your workflow
      * FACT: environment is reflected in the OIDC token and enforced by pub.dev; modifying workflow files won't work around it.
      * Suggestion to use "environment protection rules" on github, which supports "required reviewers"
      * Screenshot: github action pending review